text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+requires = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The following keys would be allowed in a feature table:
+
+-   `requires`: This is synonymous with the existing array describing required
+    features. For example, `foo = ["dep:serde", "otherfeat"]` will be identical to
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning.
```

Comment:
Should we restrict this key to only when the minimum of the version requirement is newer than `since`?  Otherwise, users can't act on it